### DISCUSSION TITLE
Attiny402/412: use TCB for fequency measurement ?

### DISCUSSION THREAD

**Discussion by BlackBrix** [SpenceKonde/megaTinyCore#1062] [Q&A]
is it possible to use the TCB for fequency measurement with this core
like described here: 
"21.3.3.1.4 Input Capture Frequency Measurement Mode"
-> https://ww1.microchip.com/downloads/en/DeviceDoc/ATtiny202-402-AVR-MCU-with-Core-Independent-Peripherals_and-picoPower-40001969A.pdf#_OPENTOPIC_TOC_PROCESSING_d114e46877
or is TCB used for different things within the core already ?

**Comment by hmeijdam** [comment]
Example

```
/*
Example for TCB0 Input Capture Frequency Measurement.
It is using pin PA2 as input and prints the frequency to the serial monitor.
If measuring low frequencies (like 50Hz), you have to select a low clock speed (e.g. 2MHz) or the 16 bit timer will overflow within one period.

Clock speed    Lowest Frequency before timer overflow
 1 MHz          15Hz 
 2 Mhz          30Hz
 4 Mhz          61Hz
 5 MHz          76Hz
 8 MHz         122Hz
10 MHz         152Hz
16 MHz         244Hz
20 MHz         305Hz

Hans Meijdam, July 2023
*/ 
volatile uint16_t timer_capture;

void setup () {
  //Serial.swap();// use this to set the TXD*/RXD* to their alternate pins
  Serial.begin(19200);// Default Serial TX is on pin 0 (PA6)
  PORTA.PIN2CTRL = PORT_PULLUPEN_bm ; // pullup on the input pin

  // configure event channel
#if defined(__AVR_TINY_2__)// The 2 series parts have a different event system configuration
  EVSYS.CHANNEL0 = EVSYS_CHANNEL0_PORTA_PIN2_gc; //link event system ch0 to pin A2
  EVSYS.USERTCB0CAPT = EVSYS_USER_CHANNEL0_gc;// Connect TCB0 capture to event channel 0
#else// This is for the 0 and 1 series parts
  EVSYS.ASYNCCH0 = EVSYS_ASYNCCH0_PORTA_PIN2_gc; //link event system async ch0 to pin A2
  EVSYS.ASYNCUSER0 = EVSYS_ASYNCUSER0_ASYNCCH0_gc;// 0x01 Link TimerB (=AYNCUSER0) to event ch0
#endif

  // configure timer B
  TCB0.CTRLB |= TCB_CNTMODE_FRQ_gc; // (0x04<<0), Input Capture Frequency Measurement
  //  on rising edge, will capture into CCMP and reset counter
  TCB0.EVCTRL |= TCB_FILTER_bm ; // activate Input Capture Noise Cancellation Filter
  TCB0.EVCTRL |= TCB_CAPTEI_bm ; // Capture Event Input Enable
//  TCB0.EVCTRL |= TCB_EDGE_bm ; // uncomment if the falling edge needs to be used. Default rising edge
  TCB0.INTCTRL |= TCB_CAPT_bm ; // Capture Interrupt Enable
  TCB0.CTRLA |= (1 << 0); // enable TCB0
}

void loop () {
  Serial.print("Timer capture: ");
  Serial.print(timer_capture);
  Serial.print("  Frequency: ");
  Serial.print(F_CPU/timer_capture);
  Serial.println(" Hz");
}

ISR (TCB0_INT_vect) {
  /*The CAPT Interrupt flag is automatically cleared 
  after the low byte of the Compare/Capture (TCBn.CCMP) register has been read.
  datasheet chapter 21.3.3.1.5  */ 
  timer_capture = TCB0_CCMP;
}
```

 TCB0 conflicts with Servo and tone() on parts that don't have TCB1
see [here](https://github.com/SpenceKonde/megaTinyCore?tab=readme-ov-file#millismicros-timekeeping-options)

**Reply by grandaspanna** [reply]
On the 412, it's possible to go a bit slower in frequency/pulse width measurement by using TCA as the clock source for TCB as TCA has more prescaler options. For the 412, by default, the core uses TCD0 as the timer for millis(). That may impact the way you do analogWrite or PWM.
![image](https://github.com/SpenceKonde/megaTinyCore/assets/54263749/5bb84a68-f647-42c9-a198-241d99b73b10)

**Comment by BlackBrix** [comment]
thanks for your very helpful answers.
prescaler would be very nice (I don't want to reduce the main clock to be able to measure low frequencies)
another thing is:
It would be nice if I could keep track of the counter overflows (with the help of an overflow interrupt),
and to be able to count into a 32bit variable (like e.g. here: https://github.com/technoblogy/frequency-probe/blob/main/FrequencyProbe.ino#L326-L335),
as far as I can see overflow interrupts are available on:
402: TCA
412: TCA, TCD
so not on TCB (?)
is that correct ?

**Reply by grandaspanna** [reply]
To get overflow interrupt in TCB, you need a 2-series device, but there's no 8-pin variant. On the 2-series, you also get the ability to piggy-back two TCB's together to get a 32-bit capture.

The DD series also has the updated TCB, but can offer up to 64k flash in a 14-pin package (as well as faster clocks at 3.3V).

The 412 is such a capable little package, but with pretty much all of the range, there's always just "one more" thing I wish they had.
